packages/rollup-core/src/app/data/consumers/l1-batch-submitter.ts

```diff
@@ -0,0 +1,311 @@
+/* External Imports */
+import {
+  getLogger,
+  logError,
+  numberToHexString,
+  remove0x,
+  ScheduledTask,
+} from '@eth-optimism/core-utils'
+import { Contract, Wallet } from 'ethers'
+
+/* Internal Imports */
+import {
+  L1BatchSubmission,
+  L2BatchStatus,
+  L2DataService,
+} from '../../../types/data'
+import { TransactionReceipt, TransactionResponse } from 'ethers/providers'
+
+const log = getLogger('l2-batch-creator')
+
+/**
+ * Polls the DB for L2 batches ready to send to L1 and submits them.
+ */
+export class L1BatchSubmitter extends ScheduledTask {
+  constructor(
+    private readonly dataService: L2DataService,
+    private readonly canonicalTransactionChain: Contract,
+    private readonly stateCommitmentChain: Contract,
+    private readonly confirmationsUntilFinal: number = 1,
+    periodMilliseconds = 10_000
+  ) {
+    super(periodMilliseconds)
+  }
+
+  /**
+   * @inheritDoc
+   *
+   * Submits L2 batches from L2 Transactions in the DB whenever there is a batch that is ready.
+   *
+   */
+  public async runTask(): Promise<void> {
+    let l2Batch: L1BatchSubmission
+    try {
+      l2Batch = await this.dataService.getNextBatchForL1Submission()
+    } catch (e) {
+      logError(log, `Error fetching batch for L1 submission! Continuing...`, e)
+      return
+    }
+
+    if (!l2Batch || !l2Batch.transactions || !l2Batch.transactions.length) {
+      log.debug(`No batches found for L1 submission.`)
+      return
+    }
+
+    let txBatchTxHash: string = l2Batch.l1TxBatchTxHash
+    let rootBatchTxHash: string = l2Batch.l1StateRootBatchTxHash
+    switch (l2Batch.status) {
+      case L2BatchStatus.BATCHED:
+        txBatchTxHash = await this.buildAndSendRollupBatchTransaction(l2Batch)
+        if (!txBatchTxHash) {
+          return
+        }
+      // Fallthrough on purpose -- this is a workflow
+      case L2BatchStatus.TXS_SUBMITTED:
+        await this.waitForTxBatchConfirms(txBatchTxHash, l2Batch.l2BatchNumber)
+      // Fallthrough on purpose -- this is a workflow
+      case L2BatchStatus.TXS_CONFIRMED:
+        rootBatchTxHash = await this.buildAndSendStateRootBatchTransaction(
```

Comment:
Talked about this with @ben-chain offline. Can update this if it's a concern, but he mentioned that @karlfloersch had been talking about potentially adding a delay for 3rd parties to append roots or some other approaches. Going to leave this as is for the time being but willing to switch if/when it makes sense.

One other thing to note is that eventually we may want to wait to submit state roots to make sure we submitted the txs at the index we expected to. After failure, the sequencer needs to catch up on the slow queue transactions and process and submit them. In that scenario, we'll probably want to submit a tx batch, wait to see our batch was added before someone appended txs from the slow queue, and then add the state roots from our batch. Otherwise we run the risk that our state root doesn't actually correspond to our tx but to a previously-submitted tx for which a state root has not yet been submitted, and therefore results in fraud. 

This race condition seems like it will always exist for state root submission, though, so maybe we always add an `expected state root queue index` such that state roots only get appended if the cumulative number of roots queued is that number 🤷

Reply:
One other thing I didn't think about is that it's very possible that we'll want to submit roots without txs. In the slow queue case, for example, we'd have a tx we have to include, but a root may not have been appended to the State Chain for it, so we end up submitting the root but not the tx so that we can then move on to submit subsequent batches.